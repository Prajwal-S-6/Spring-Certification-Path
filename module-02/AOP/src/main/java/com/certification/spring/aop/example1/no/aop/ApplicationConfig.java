package com.certification.spring.aop.example1.no.aop;

import org.springframework.context.annotation.ComponentScan;

@ComponentScan
public class ApplicationConfig {
}
