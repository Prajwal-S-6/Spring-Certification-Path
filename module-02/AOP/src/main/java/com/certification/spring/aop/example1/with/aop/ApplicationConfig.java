package com.certification.spring.aop.example1.with.aop;

import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.EnableAspectJAutoProxy;

@ComponentScan
@EnableAspectJAutoProxy
public class ApplicationConfig {
}
