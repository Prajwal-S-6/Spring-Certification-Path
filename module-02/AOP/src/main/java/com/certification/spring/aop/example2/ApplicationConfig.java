package com.certification.spring.aop.example2;

import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.EnableAspectJAutoProxy;

@ComponentScan
@EnableAspectJAutoProxy
public class ApplicationConfig {
}
