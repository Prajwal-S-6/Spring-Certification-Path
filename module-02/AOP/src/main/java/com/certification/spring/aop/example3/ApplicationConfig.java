package com.certification.spring.aop.example3;

import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.EnableAspectJAutoProxy;

@ComponentScan
///  enables CGLIB proxy even though the bean implements the interface
@EnableAspectJAutoProxy(proxyTargetClass = true)
// @EnableAspectJAutoProxy
public class ApplicationConfig {
}
