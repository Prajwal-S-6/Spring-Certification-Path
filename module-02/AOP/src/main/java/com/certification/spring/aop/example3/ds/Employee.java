package com.certification.spring.aop.example3.ds;

public class Employee {
}
