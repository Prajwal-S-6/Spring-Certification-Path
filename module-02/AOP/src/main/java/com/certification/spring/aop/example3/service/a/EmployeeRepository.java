package com.certification.spring.aop.example3.service.a;


import com.certification.spring.aop.example3.ds.Employee;

public interface EmployeeRepository {
    Employee findEmployeeById(long id);

    void saveEmployee(Employee employee);

    void deleteEmployee(Employee employee);

}
