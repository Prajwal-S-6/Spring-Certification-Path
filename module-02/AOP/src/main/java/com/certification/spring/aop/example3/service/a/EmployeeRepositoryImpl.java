package com.certification.spring.aop.example3.service.a;

import com.certification.spring.aop.example3.ds.Employee;
import org.springframework.stereotype.Component;

@Component
public class EmployeeRepositoryImpl implements EmployeeRepository {
    @Override
    public Employee findEmployeeById(long id) {
        return new Employee();
    }

    @Override
    public void saveEmployee(Employee employee) {
    }

    @Override
    public void deleteEmployee(Employee employee) {
    }
}
