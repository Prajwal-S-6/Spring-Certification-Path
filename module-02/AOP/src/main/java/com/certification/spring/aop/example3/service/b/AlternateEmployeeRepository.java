package com.certification.spring.aop.example3.service.b;


import com.certification.spring.aop.example3.ds.Employee;
import org.springframework.stereotype.Component;

@Component
public class AlternateEmployeeRepository {
    public Employee findEmployeeById(long id) {
        return new Employee();
    }

    public void saveEmployee(Employee employee) {
    }

    public void deleteEmployee(Employee employee) {
    }
}
