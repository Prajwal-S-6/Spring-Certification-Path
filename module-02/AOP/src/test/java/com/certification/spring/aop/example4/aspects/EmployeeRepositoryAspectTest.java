package com.certification.spring.aop.example4.aspects;

import com.certification.spring.aop.example4.ApplicationConfig;
import com.certification.spring.aop.example4.ds.Employee;
import com.certification.spring.aop.example4.service.a.EmployeeRepository;
import com.certification.spring.aop.example4.service.a.EmployeeRepositoryImpl;
import com.certification.spring.aop.example4.service.b.AlternateEmployeeRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.util.ReflectionTestUtils;

import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.*;

@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = {ApplicationConfig.class})
class EmployeeRepositoryAspectTest {

    @Autowired
    EmployeeRepository employeeRepository;

    @Autowired
    AlternateEmployeeRepository alternateEmployeeRepository;


    @Test
    void shouldNotLogForFinalMethods() {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        PrintStream originalOutput = System.out;
        System.setOut(new PrintStream(byteArrayOutputStream));

        alternateEmployeeRepository.deleteEmployee(new Employee());
        alternateEmployeeRepository.saveEmployee(new Employee());
        alternateEmployeeRepository.findEmployeeById(1L);
        alternateEmployeeRepository.findAndUpdateEmployeeById(1L, new Employee());

        System.setOut(originalOutput);
        String logMessage = byteArrayOutputStream.toString();

        assertThat(logMessage).contains(List.of("Before - findAndUpdateEmployeeById", "After - findAndUpdateEmployeeById"));

    }

    @Test
    void shouldNotLogForSelfInvokingMethods() throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        PrintStream originalOutput = System.out;
        System.setOut(new PrintStream(byteArrayOutputStream));

        employeeRepository.saveEmployee(new Employee());
        employeeRepository.deleteEmployee(new Employee());
        employeeRepository.findAndUpdateEmployeeById(1L , new Employee());


        System.setOut(originalOutput);
        String logMessage = byteArrayOutputStream.toString();

        assertThat(logMessage).contains(List.of("Before - saveEmployee",
                "Before - deleteEmployee",
                "Before - findAndUpdateEmployeeById",
                "After - saveEmployee",
                "After - deleteEmployee",
                "After - findAndUpdateEmployeeById"));

        assertThat(logMessage).doesNotContain("Before - findEmployeeById",
                "After - findEmployeeById", "Before - deleteByEmail", "After - deleteByEmail");
    }

}