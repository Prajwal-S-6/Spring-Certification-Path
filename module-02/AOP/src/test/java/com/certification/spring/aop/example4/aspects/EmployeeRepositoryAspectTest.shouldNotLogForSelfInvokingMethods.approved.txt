[INFO] Before -- saveEmployee,
[INFO] After -- saveEmployee,
[INFO] Before -- deleteEmployee,
[INFO] After -- deleteEmployee,
[INFO] Before -- findAndUpdateEmployeeById,
[INFO] After -- findAndUpdateEmployeeById